zcash_client_sqlite/src/wallet.rs

```diff
@@ -461,30 +515,78 @@ pub fn rewind_to_height<P: consensus::Parameters>(
             })?;
 
-    // nothing to do if we're deleting back down to the max height
-    if block_height >= last_scanned_height {
-        Ok(())
+    let rewind_height = if block_height >= (last_scanned_height - PRUNING_HEIGHT) {
+        Some(block_height)
     } else {
-        // Decrement witnesses.
-        wdb.conn.execute(
-            "DELETE FROM sapling_witnesses WHERE block > ?",
-            &[u32::from(block_height)],
-        )?;
-
-        // Un-mine transactions.
-        wdb.conn.execute(
-            "UPDATE transactions SET block = NULL, tx_index = NULL WHERE block > ?",
-            &[u32::from(block_height)],
-        )?;
+        match get_rewind_height(&wdb)? {
+            Some(h) => {
+                if block_height > h {
+                    return Err(SqliteClientError::RequestedRewindInvalid(h));
+                } else {
+                    Some(block_height)
+                }
+            }
+            None => Some(block_height),
+        }
+    };
 
-        // Now that they aren't depended on, delete scanned blocks.
-        wdb.conn.execute(
-            "DELETE FROM blocks WHERE height > ?",
-            &[u32::from(block_height)],
-        )?;
+    // nothing to do if we're deleting back down to the max height
 
-        Ok(())
+    if let Some(block_height) = rewind_height {
+        if block_height < last_scanned_height {
+            // Decrement witnesses.
+            wdb.conn.execute(
+                "DELETE FROM sapling_witnesses WHERE block > ?",
+                &[u32::from(block_height)],
+            )?;
+
+            // Un-mine transactions.
+            wdb.conn.execute(
+                "UPDATE transactions SET block = NULL, tx_index = NULL WHERE block > ?",
+                &[u32::from(block_height)],
+            )?;
+
+            // Now that they aren't depended on, delete scanned blocks.
+            wdb.conn.execute(
+                "DELETE FROM blocks WHERE height > ?",
+                &[u32::from(block_height)],
+            )?;
+
+            // Rewind received notes
+            wdb.conn.execute(
+                "DELETE FROM received_notes
+                    WHERE id_note IN (
+                        SELECT rn.id_note
+                        FROM received_notes rn
+                        LEFT OUTER JOIN transactions tx
+                        ON tx.id_tx = rn.tx
+                        WHERE tx.block > ?
+                    );",
+                &[u32::from(block_height)],
+            )?;
+
+            // Rewind sent notes
+            wdb.conn.execute(
+                "DELETE FROM sent_notes
+                    WHERE id_note IN (
+                        SELECT sn.id_note
+                        FROM sent_notes sn
+                        LEFT OUTER JOIN transactions tx
+                        ON tx.id_tx = sn.tx
+                        WHERE tx.block > ?
+                    );",
+                &[u32::from(block_height)],
+            )?;
```

Comment:
Bug: This attempts to rewind notes based on the block their transaction was mined in, _after_ the transactions they are connected to were unmined in the database. As a result, this will always do nothing (because `NULL` [sorts as less than everything else in SQLite](https://www.sqlite.org/datatype3.html#sort_order)).

In general, the rewind logic needs to be executed in the opposite order to how the data is stored. I haven't checked this, but I suspect the order needs to be:
- Decrement witnesses.
- Rewind received notes.
- Rewind sent notes.
- Rewind UTXOs.
  - It currently has no links to its containing transaction, but if we altered their table to link to the `transactions` table, we'd need to rewind it in the same place as received notes.
- Un-mine transactions.
- Delete scanned blocks.

I'm surprised that this bug wasn't noticed yet, given that this logic was added by the wallet team in https://github.com/zcash/librustzcash/commit/7fc2d9725c2ff0be6f5a63d0ec1cfeb72e2f1b65 I presume to fix a specific problem they were having. The first question that comes to mind is, did they _actually_ want the note data deleted? I vaguely recall deciding not to rewind this data when I first wrote this function because I wanted to preserve local history (in the case of transactions that never got mined), relying on the "mined transaction" unlinking to indicate the transaction wasn't contributing to balance. Right now, notes from rewound txs aren't being deleted but UTXOs are (because their table wasn't hooked up to the `transactions` table); was it actually just the UTXOs that the wallet team were trying to clear out? cc @pacu

Reply:
I was going through the old issues, and rediscovered https://github.com/zcash/librustzcash/issues/208:
> We have isolated the problem. It occurs when we are trying to `insert` a received note pointing to a transaction that already exists, rather than `update` it. The first fix is to also delete from the `received_notes` table during a rollback, which prevents this situation from occurring. The next (optional) fix is to also change the `insert` to an `update` when appropriate, which would avoid the foreign key constraint problem if the wallet ever gets in this state, somehow.

I suspect this is what motivated these changes, which for a long time lived in Kevin's local forks of `zcash/librustzcash`.

Reply:
So if that was indeed the motivation, then what we should instead be doing is the second half of the above comment: checking if the received (or sent) note exists, and updating it if so rather than trying to insert assuming it doesn't exist. This would fit better with the model of "don't delete user-created data that can't be recovered from the chain", which is the case for created transactions that then never get mined (which would have `sent_notes` outputs as well as `received_notes` change outputs).

Reply:
There were many problems with the rewind APi. One was the constraints issue, the other one was the not being able to rewind past 100 blocks. When that happened the wallet would lose ability to spend funds https://github.com/zcash/ZcashLightClientKit/issues/270
https://github.com/zcash/librustzcash/issues/373

I don't know if that's what you. were referring to.

Reply:
The "rewind past 100 blocks" issue is distinct from this API issue; that problem is due to the number of witnesses maintained (and that limit is becoming even more strict in zebrad). If you need to rewind that far, there's nothing you can do except start resyncing again from essentially scratch.

Reply:
We do provide the `put_sent_note` and `put_received_note` methods which have exactly the "update-or-insert" behavior. But these are now deprecated and should no longer be considered part of the public api; they're purely used internally in the implementation of `store_sent_tx` and `store_decrypted_tx`.

Reply:
I'm torn here. I really prefer the delete-then-reinsert approach, unless we're going to make maintaining unmined transaction data a first-class citizen of this API. I don't think we're going to do that now, so I think that we should correct the order of operations in the deletion that @str4d identified, but otherwise keep this as it is. Then, we can switch to an implementation that is safer with respect to history preservation when we make that a first-class concern.